common/djangoapps/third_party_auth/lti.py

```diff
@@ -0,0 +1,142 @@
+
+        # Save validated LTI parameters (or None if invalid or not submitted)
+        validated_lti_params = self.get_validated_lti_params(self.strategy)
+        if not validated_lti_params:
+            self.strategy.session_set(LTI_PARAMS_KEY, None)
+            raise AuthFailed(self)
+        else:
+            logout(self.strategy.request)
+            self.strategy.session_set(LTI_PARAMS_KEY, validated_lti_params)
+
+        # Save extra data into session.
+        # While Basic LTI 1.0 specifies that the message is to be signed using OAuth, implying
+        # that any GET parameters should be stripped from the base URL and included as signed
+        # parameters, typical LTI Tool Consumer implementations do not support this behaviour. As
+        # a workaround, we accept TPA parameters from LTI custom parameters prefixed with "tpa_".
+
+        for field_name in self.setting('FIELDS_STORED_IN_SESSION', []):
+            if 'custom_tpa_' + field_name in validated_lti_params:
+                self.strategy.session_set(field_name, validated_lti_params['custom_tpa_' + field_name])
+
+        if 'custom_tpa_' + redirect_name in validated_lti_params:
+            # Check and sanitize a user-defined GET/POST next field value
+            redirect_uri = validated_lti_params['custom_tpa_' + redirect_name]
+            if self.setting('SANITIZE_REDIRECTS', True):
+                redirect_uri = sanitize_redirect(self.strategy.request_host(), redirect_uri)
+            self.strategy.session_set(redirect_name, redirect_uri or self.setting('LOGIN_REDIRECT_URL'))
+
+        # Set a auth_entry here so we don't have to receive that as a custom parameter
+        self.strategy.session_setdefault('auth_entry', 'login')
+
+    def auth_complete(self, *args, **kwargs):
+        lti_params = self.strategy.session_get(LTI_PARAMS_KEY)
+        kwargs.update({'response': {LTI_PARAMS_KEY: lti_params}, 'backend': self})
+        return self.strategy.authenticate(*args, **kwargs)
+
+    def get_user_id(self, details, response):
+        lti_params = response[LTI_PARAMS_KEY]
+        return lti_params['oauth_consumer_key'] + ":" + lti_params['user_id']
+
+    def get_user_details(self, response):
+        details = {}
+        lti_params = response[LTI_PARAMS_KEY]
+
+        def add_if_exists(lti_key, details_key):
+            if lti_key in lti_params and lti_params[lti_key]:
+                details.update({details_key: lti_params[lti_key]})
+
+        add_if_exists('email', 'email')
+        add_if_exists('lis_person_name_full', 'fullname')
+        add_if_exists('lis_person_name_given', 'first_name')
+        add_if_exists('lis_person_name_family', 'last_name')
+        return details
+
+    @classmethod
+    def get_validated_lti_params(cls, strategy):
+        request = Request(uri=strategy.request.build_absolute_uri(),
+                          http_method=strategy.request.method, body=strategy.request.body)
```

Comment:
Our style guidelines request you break calls like this, to prevent code squashed to the right:
```python
request = Request(
    uri=strategy.request.build_absolute_uri(),
    http_method=strategy.request.method, body=strategy.request.body
)
```

Please see https://github.com/edx/edx-platform/wiki/Python-Guidelines#breaking-long-lines